Avoid converting Quat to Euler angles when not necessary.

tagcup opened the pull request:
Also ensure that get_scale doesn't arbitrarlity change the signs of scales, ensuring that the combination of get_rotation and get_scale gives the correct basis.

Added various missing functions and constructors.

Should close #17968.

tagcup committed:
Avoid converting Quat to Euler angles when not necessary.

Also ensure that get_scale doesn't arbitrarlity change the signs of scales, ensuring that the combination of get_rotation and get_scale gives the correct basis.

Added various missing functions and constructors.

Should close #17968.

akien-mga commented on core/math/matrix3.h:
```diff
@@ -219,6 +230,8 @@ class Basis {
 	Basis(const Quat &p_quat) { set_quat(p_quat); };
 	Basis(const Vector3 &p_euler) { set_euler(p_euler); }
 	Basis(const Vector3 &p_axis, real_t p_phi) { set_axis_angle(p_axis, p_phi); }
+	Basis(const Vector3 &p_axis, real_t p_phi, const Vector3 &p_scale) { set_axis_angle_scale(p_axis, p_phi, p_scale); }
```
Are those new constructors intended to be exposed to GDScript (currently they wouldn't be, see #13151)?

akien-mga replied:
If not it's good as is I think.

akien-mga commented:
Thanks! (checked by @reduz who confirms it's a great addition)

tagcup commented:
The API for (object-)local operations isn't on par with the rest of the API, I can do another PR for that if there's interest for those functions (they'd for sure be relevant for FPS "eyes" or direct camera control situations where the object-local frame makes sense)

hpvb commented:
I'm a little worried about this one as it changes the behavior of get_scale() and replaces it with a different function called get_scale_abs() for 3.0 I'd like to avoid something like this.

If we want this fix would it make sense to make sure that get_scale() in gdscript and other bindings get forwarded to the new get_scale_abs() with a new function for get_scale_rotated() (or similar)?

Not pulling this in 3.0.3 for now, but keeping the tag for 3.0.4 and see if we can discuss the above.

tagcup commented:
There's nothing to discuss there. That's how get_scale works. Someone, probably Odino, changed it at some point, adding an ad-hoc and and moved the original to a function with _signed prefix, because it was apparently convenient for bullet bindings. This restores the old and correct behavior. Having an arbitrary abs there is a bug: you won't get the the original basis when you use it along with get rotate.

akien-mga commented:
Now that Godot 3.1 has been released, we only plan to cherry-pick critical bug fixes to the 3.0 branch, unless there is very strong support in favor of a specific non-critical issue. Removing cherrypick label for 3.0.
